design/cdk-assets.md

```diff
@@ -0,0 +1,244 @@
+found    aws-cdk-images-3333333333EU-eu-west-2:d31ca1aef8d1b68217852e7aea70b1e857d107b47637d5160f9f9a1b24882d2a
+done     d31ca1aef8d1b68217852e7aea70b1e857d107b47637d5160f9f9a1b24882d2a    
+--------------------------------------------------------------------------
+asset    a0bae29e7b47044a66819606c65d26a92b1e844f4b3124a5539efc0167a09e57
+found    s3://aws-cdk-files-2222222222US-us-east-1/a0bae29e7b47044a66819606c65d26a92b1e844f4b3124a5539efc0167a09e57.zip
+assume   arn:aws:iam::3333333333EU:role/aws-cdk-publish-3333333333EU-eu-west-2
+notfound s3://aws-cdk-files-3333333333EU-us-west-2/a0bae29e7b47044a66819606c65d26a92b1e844f4b3124a5539efc0167a09e57.zip
+cached   zip ./myzipdirectory
+upload   s3://aws-cdk-files-3333333333EU-us-west-2/a0bae29e7b47044a66819606c65d26a92b1e844f4b3124a5539efc0167a09e57.zip
+done     a0bae29e7b47044a66819606c65d26a92b1e844f4b3124a5539efc0167a09e57
+--------------------------------------------------------------------------
+```
+
+The log above describes the following:
+
+The first asset to process is the docker image with id `d31ca1a...`. We first assume the role is the 2222US environment and check if the image exists in this ECR repository. Since it doesn't exist (`notfound`), we check if it exists in the local cache under the tag `d31ca1aef8d1b68217852e7aea70b1e857d107b47637d5160f9f9a1b24882d2a`. It doesn't so we build the docker image (`package`) and then push it. Then we assume the role from the 33333EU environment and find that the image is already in that ECR repository, so we just finish.
+
+The second asset is the file asset with id `a0bae...`. The first environment doesn't specify a role, so we just check if the file exists in the S3 bucket. It is, so we move to the next destination. We assume the role and check if the asset is in the s3 location. It is not (`notfound`), so we need to package and upload it. Before we package we check if it's cached locally and find that it is (`cached`), so no need to zip again, just `upload`.
+
+### `ls`
+
+```shell
+cdk-assets ls DIR
+```
+
+Prints a list of asset identifiers and their type.
+
+Example:
+
+```shell
+$ cdk-assets ls cdk.out
+d31ca1aef8d1b68217852e7aea70b1e857d107b47637d5160f9f9a1b24882d2a image
+a0bae29e7b47044a66819606c65d26a92b1e844f4b3124a5539efc0167a09e57 file
+```
+
+### Programmatic API
+
+The tool should expose a programmatic (library) API, so it can be integrated with other tools such as the AWS CLI and IDEs. The library should be jsii-compliant and released to all languages supported by the AWS CDK.
+
+Since the publishing process is highly asynchronous, the API should include the ability to subscribe to progress events in order to allow implementation of a rich user interface.
+
+Proposed API:
+
+```ts
+class Assets {
+  constructor(dir: string);
+  readonly manifest: AssetManifest;
+  
+  // starts publishing a single asset
+  publish(assetid: string, progress?: Progress): Progress;
```

Comment:
I think the return type you want here is `Publish`.

Reply:
Yap, you are right.

Reply:
Of course.